### PR TITLE
new shared project (OpenTelemetry.Shared) and move Guard.cs

### PR DESCRIPTION
**Pull request by TimothyMothra**
Towards #4348.

## Background
I'm working to create a single shared project for common internal classes.
This will hold internal classes meant to be referenced by libraries in this repo.
This should eliminate the need for InternalsVisableTo and prevent conflicts when expanding internal classes.

This work was inspired by the Contrib repo, however the Contrib repo has two shared projects ([OpenTelemetry.Contrib.Shared](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Contrib.Shared) and [OpenTelemetry.Internal](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Internal)).

## Changes

Please provide a brief description of the changes here.

- new project `OpenTelemetry.Shared`. 
  Shared files from OpenTelemetry.csproj or OpenTelemetry.Api.csproj will be moved to this project in a future PR.
- create subdirectory `Internals`.
  This will hold files intended to be included in all other projects.

- move `Guard.cs` from OpenTelemetry.Api to OpenTelemetry.Shared\Internals.
- split `Guard.cs` into three files, one for each class: `Guard`, `NullableAttributes`, and `CallerArgumentExpressionAttribute`

- update other projects with references to `Guard.cs`.
  

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Api/OpenTelemetry.Api.csproj [hunk 1]
```diff
@@ -15,4 +15,9 @@
   <ItemGroup>
     <PackageReference Include="System.Diagnostics.DiagnosticSource" Version="$(SystemDiagnosticSourcePkgVer)" />
   </ItemGroup>
+
+  <ItemGroup>
```
Comment (utpilla) [inline src/OpenTelemetry.Api/OpenTelemetry.Api.csproj:19]: Include these files for the SDK and ProviderBuilderExtensions project as well?

**Comment by utpilla** [comment]
> I'm working to create a single shared project for common internal classes.
This will hold internal classes meant to be referenced by libraries in this repo.
**This should eliminate the need for InternalsVisibleTo and prevent conflicts when expanding internal classes.**

Is the `InternalsVisibleTo` only required for the `Guard` class? The SDK also uses a few other API internal stuff as well such as the setter for `Propagators.DefaultTextMapPropagator`: https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry/Sdk.cs#L58 

I don't think we would want to extract out `Propagators` class just to get rid of the `InternalsVisibleTo`.

**Review by reyang — CHANGES_REQUESTED** [review]
Need to discuss about the direction.

**Comment by reyang** [comment]
> * create subdirectory `Internals`.
>   This will hold files intended to be included in all other projects.

Why is such folder needed?

> * split `Guard.cs` into three files, one for each class: `Guard`, `NullableAttributes`, and `CallerArgumentExpressionAttribute`

Why is this split needed?

**Comment by reyang** [comment]
> I don't think we would want to extract out `Propagators` class just to get rid of the `InternalsVisibleTo`.

+1

**Pull request closed without merging**